### PR TITLE
dtls.c: Check for optional handshake messages

### PR DESCRIPTION
**Pull request by boaks**
In general the handshake state machine forces the handshakes to go
through
a defined order. But optional handshake message are not controlled by
that.

This fix introduces a optional handshake message field. With that, the
state-machine is enabled to check, if this message is (optionally)
expected. When processed, the field is reset and receiving the optional
message twice will therefore fail the handshake.

Signed-off-by: Achim Kraus <achim.kraus@bosch.io>

### CODE REVIEW

**Comment by boaks** [comment]
Alternative proposal for #88

**Comment by boaks** [comment]
This one would be my favorite for the next merge :-).
It fixes #78 and obsoletes #88.

**Comment by boaks** [comment]
@obgm

That 20 lines would close three topics, #78, #88 and this PR itself :-).

File: peer.h [hunk 1]
```diff
@@ -52,6 +57,7 @@ typedef struct dtls_peer_t {
 
   dtls_peer_type role;       /**< denotes if this host is DTLS_CLIENT or DTLS_SERVER */
   dtls_state_t state;        /**< DTLS engine state */
+  int16_t optionalHandshakeMessage; /**< optional next handshake message, DTLS_HT_NO_OPTIONAL_MESSAGE, if no optional message is expected. */
```
Comment (obgm) [inline peer.h:60]: Please follow C coding conventions. An identifier thus would merely look like this: `optional_handshake_message`.

File: dtls.c [hunk 2]
```diff
@@ -3628,10 +3633,11 @@ handle_handshake_msg(dtls_context_t *ctx, dtls_peer_t *peer, uint8 *data, size_t
 #ifdef DTLS_ECC
   case DTLS_HT_CERTIFICATE_REQUEST:
 
-    if (state != DTLS_STATE_WAIT_SERVERHELLODONE) {
+    if (!is_tls_ecdhe_ecdsa_with_aes_128_ccm_8(peer->handshake_params->cipher) ||
```
Comment (obgm) [inline dtls.c:3636]: Hm, the newly introduced first check seems to be unrelated to this PR. This might warrant at least a separate commit.

**Review by obgm — CHANGES_REQUESTED** [review]
Thank you for this fix. I wonder why `DTLS_HT_NO_OPTIONAL_MESSAGE` is defined in *peer.h* rather than in *dtls.c* which is the only place where it is used. (The other `DTLS_HT_` types are defined in _dtls.h_.)
Moreover, as the PR description indicates, this flag directly relates to the handshake state. Therefore, I think it should go into the `dtls_handshake_parameters_t` structure (possibly even as an enum?).

I also suggest some minor changes to the commit message: First, lines longer than 72 for characters are rather unusual and lead to the stray line breaks as seen in the GitHub presentation. Second, I suggest slight rewording to avoid too much passive tense, e.g., as shown below (with punctuation error fixed):

```
dtls.c: Check for optional handshake messages

In general the handshake state machine forces the handshakes to go
through a defined order. But optional handshake message are not
controlled by that.

This fix introduces a optional handshake message field. This new
field enables the state-machine to check if this message
is (optionally) expected. When processed, the field is reset and
receiving the optional message twice will therefore fail the
handshake.


Fixes: Issue #78

Signed-off-by: Achim Kraus <achim.kraus@bosch.io>
```

File: peer.h (hunk 1 above)
Comment (boaks) [reply peer.h:60]: Too much java ...

File: dtls.c (hunk 2 above)
Comment (boaks) [reply dtls.c:3636]: If you prefer it, sure.

**Comment by boaks** [comment]
> I wonder why DTLS_HT_NO_OPTIONAL_MESSAGE is defined in peer.h rather than in dtls.c which is the only place where it is used. (The other DTLS_HT_ types are defined in dtls.h.)

My intention was to define it close to the field, where the value is used. I can move it to dtls.h.

**Comment by boaks** [comment]
Next try. 
(If the two empty lines before "Fixes: Issue " intended? I reduced it to one, but if it's intended, I can add a additional one.)

**Comment by obgm** [comment]
> Next try. (Is the two empty lines before "Fixes: Issue " intended? I reduced it to one, but if it's intended, I can add a additional one.)

Sorry, this was a copy&paste bug. Thanks for fixing this right away.

**Comment by obgm** [comment]
Thanks for the changes which look good so far.
The only open question now is whether the new flag `optional_handshake_message` should be in the `dtls_handshake_parameters_t` structure?

**Comment by boaks** [comment]
> The only open question now is whether the new flag optional_handshake_message should be in the dtls_handshake_parameters_t structure?

My intention to put it in `dtls_peer_t` was to keep it close to `dtls_state_t state`, though for me both are related.
Anyway, if you prefer, it could be moved to `dtls_handshake_parameters_t`.

**Comment by obgm** [comment]
Agreed: Keeping it close to state may make even more sense than close to hs_state. My bad.

**Review by obgm — APPROVED** [review]

**Comment by boaks** [comment]
I considered the other variant as well. It saves 2 bytes per peer. In my opinion, it's not that easy to understand, why that is required and working, so I gone for have it close to the related state.

Thanks for merging.
